### PR TITLE
Add parsing anchor table format 2

### LINKED ISSUE #237: anchorFormat 2 not supported. Should be '1'.
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of Fonts
- [ ] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/SixLabors/Fonts/issues) and [closed](https://github.com/SixLabors/Fonts/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
After upgrading from beta15 to beta16 (via SixLabors.ImageSharp.Drawing), I often see this error in my log files.
For now I have rolled back to beta15. I noticed that the font API has changed slightly in beta16, so maybe this is now intended but my code is no longer using fonts correctly?

Any pointers as to what I might be doing wrong or what I can do to work around it would be appreciated.

> anchorFormat 2 not supported. Should be '1'.

```
System.NotSupportedException:
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPos.AnchorTable.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPos.MarkArrayTable..ctor (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPos.LookupType4SubTable+LookupType4Format1SubTable.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPos.LookupType4SubTable.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPos.LookupTable.LoadLookupSubTable (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPos.LookupTable.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPos.LookupListTable.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPosTable.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.AdvancedTypographic.GPosTable.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.Tables.TableLoader.Load (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.FontReader.TryGetTable (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.StreamFontMetrics.LoadFont (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.StreamFontMetrics.LoadFont (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.StreamFontMetrics.LoadFont (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.FileFontMetrics+<>c__DisplayClass3_0.<.ctor>b__0 (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at System.Lazy`1.ViaFactory (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Lazy`1.CreateValue (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at SixLabors.Fonts.FileFontMetrics.TryGetGlyphId (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.TextLayout.DoFontRun (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.TextLayout.ProcessText (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at SixLabors.Fonts.TextLayout.GenerateLayout (SixLabors.Fonts, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13)
   at ...userland...
```

### Steps to Reproduce
```csharp
var fontFamily = default(FontFamily);
if (!SystemFonts.TryGet("DejaVu Sans", out fontFamily))
{
    throw new Exception($"Unable to find a suitable font. Available options are: {String.Join(", ", SystemFonts.Families.Select(x => x.Name))}");
}
var badgeFont = new Font(fontFamily, 12, FontStyle.Regular);
var badgeTextSize = TextMeasurer.Measure("100", new TextOptions(badgeFont));
```

### System Configuration

- Fonts version: beta16
- Other Six Labors packages and versions: ImageSharp 2.0, ImageSharp.Drawing 2.0
- Environment (Operating system, version and so on): Linux
- .NET Framework version: 6.0
- Additional information: Azure App Service

### PR DESCRIPTION
**Pull request by brianpopow**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/Fonts/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [ ] I have provided test coverage for my change (where applicable)

### Description

This adds parsing [Anchor table format 2](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#anchorTables).
The difference to format 1 is, that there is a additional Index to glyph contour point, which should be used for hinting.

We dont support hinting yet, right? The output looks correct from what I can tell without it.

closes #237

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
>We dont support hinting yet, right? 

We actually do, it's hidden behind a property in the `TextOptions` type.

I think the provided index represents the index of the point in the `GlyphOutline.ControlPoints`.

I had a quick look at `hb-ot-layout-gpos-table.hh` line 426-434 and I _think_ we should be returning that point if the index is viable and we have hinting enabled. I could be horribly wrong though as I found it impossible to actually navigate through the various levels of indirection to see exactly what `get_glyph_contour_point_for_origin` does.

**Comment by brianpopow** [comment]
> I had a quick look at hb-ot-layout-gpos-table.hh line 426-434 and I think we should be returning that point if the index is viable and we have hinting enabled. I could be horribly wrong though as I found it impossible to actually navigate through the various levels of indirection to see exactly what get_glyph_contour_point_for_origin does.

This looks about right. To really get this correct, I think we need a real world example where this feature is actually used. The one example from [microsoft docs](https://docs.microsoft.com/en-us/typography/opentype/spec/) does not do it for me. I wish they would provide fonts along those examples.

**Comment by JimBobSquarePants** [comment]
> I wish they would provide fonts along those examples.

Completely agree, the savings in development time would be incredible!

I think we should go ahead and grab the point from the contour points when possible. I'd have to double check to see if we should be scaling the value like Harfbuzz does. I'm surprised they don't have a test font squirreled away somewhere.

File: src/SixLabors.Fonts/Tables/AdvancedTypographic/GPos/AnchorTable.cs [hunk 1]
```diff
@@ -45,7 +45,8 @@ public static AnchorTable Load(BigEndianBinaryReader reader, long offset)
             return anchorFormat switch
             {
                 1 => AnchorFormat1.Load(reader),
-                _ => throw new NotSupportedException($"anchorFormat {anchorFormat} not supported. Should be '1'.")
+                2 => AnchorFormat2.Load(reader),
```
Comment (rhyskoedijk) [inline src/SixLabors.Fonts/Tables/AdvancedTypographic/GPos/AnchorTable.cs:48]: As somebody who knows nothing about fonts, making these error messages more descriptive might be helpful too. If I knew that my font format wasn't supported, I'd just pick a different one.

```csharp
 3 => throw new NotSupportedException($"Variable fonts (format 3) are not supported. Read: https://docs.sixlabors.com/articles/fonts/anchorTableFormats.html")
 _ => throw new NotSupportedException($"Fonts using anchor table format {anchorFormat} are not supported. Read: https://docs.sixlabors.com/articles/fonts/anchorTableFormats.html")
```

Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Tables/AdvancedTypographic/GPos/AnchorTable.cs:48]: I wouldn't link to docs from exception messages. The maintenance burden of tracking URL changes would be too much.

Comment (rhyskoedijk) [reply src/SixLabors.Fonts/Tables/AdvancedTypographic/GPos/AnchorTable.cs:48]: that's fair enough, was just a thought

**Comment by JimBobSquarePants** [comment]
@brianpopow I decided to actually go ahead and implement anchor format 2 table as I'm fairly confident of the approach. I've also added the font to the solution so we can test it. 

I've been trying various character combinations to see if I can trigger the table but have so far failed.

**Comment by brianpopow** [comment]
> I've been trying various character combinations to see if I can trigger the table but have so far failed.

I tried to look through the harfbuzz tests for examples, but so far also without luck. I am pretty sure there must be an example there, but dont know how to find one.

**Comment by JimBobSquarePants** [comment]
>I tried to look through the harfbuzz tests for examples, but so far also without luck. I am pretty sure there must be an example there, but dont know how to find one.

@behdad @khaledhosny Would you possibly be able to help here? We're looking for example fonts and tests that allow us to test features using AnchorTable Formats 2 and 3.

**Comment by behdad** [comment]
> @behdad @khaledhosny Would you possibly be able to help here? We're looking for example fonts and tests that allow us to test features using AnchorTable Formats 2 and 3.

I'm not aware of any fonts using those currently.

**Comment by JimBobSquarePants** [comment]
Thanks @behdad at least this means if we make a mistake, no one will see it 😂

**Comment by JimBobSquarePants** [comment]
@brianpopow I'm gonna merge this as-is since we can't provide a test font. I'll add an issue to flesh out `AnchorFormat3` to add Device table support.

**Review by JimBobSquarePants — APPROVED** [review]

**Pull request merged**